tools/ar_util.py

```diff
@@ -0,0 +1,230 @@
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import os
+import re
+import hashlib
+import functools
+import pickle
+
+from elftools.elf import elffile
+from collections import defaultdict
+
+try:
+    from ar import Archive
+except:
+    Archive = None
+
+
+class PickleCache:
+    def __init__(self, path, prefix=""):
+        self.path = path
+        self._get_fn = lambda key: os.path.join(path, prefix + key[:24])
+
+    def store(self, key, data):
+        os.makedirs(self.path, exist_ok=True)
+        with open(self._get_fn(key), "wb") as f:
+            pickle.dump(data, f)
+
+    def load(self, key):
+        with open(self._get_fn(key), "rb") as f:
+            return pickle.load(f)
+
+
+def cached(key, cache):
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            cache_key = key(*args, **kwargs)
+            try:
+                d = cache.load(cache_key)
+                if d["key"] != cache_key:
+                    raise Exception("Cache key mismatch")
+                return d["data"]
+            except Exception:
+                res = func(*args, **kwargs)
+                try:
+                    cache.store(
+                        cache_key,
+                        {
+                            "key": cache_key,
+                            "data": res,
+                        },
+                    )
+                except Exception:
+                    pass
+                return res
+
+        return wrapper
+
+    return decorator
+
+
+class CachedArFile:
+    def __init__(self, fn):
+        if not Archive:
+            raise RuntimeError("Please run 'pip install ar' to link .a files")
+        self.fn = fn
+        self._archive = Archive(open(fn, "rb"))
+        info = self.load_symbols()
+        self.objs = info["objs"]
+        self.symbols = info["symbols"]
+
+    def open(self, obj):
+        return self._archive.open(obj, "rb")
+
+    def _cache_key(self):
+        sha = hashlib.sha256()
+        with open(self.fn, "rb") as f:
+            for chunk in iter(lambda: f.read(4096), b""):
+                sha.update(chunk)
+        # Change this salt if the cache data format changes
+        sha.update(bytes.fromhex("45155db4bc868fa78cb99c3448b2bf2b"))
```

Comment:
This is something of a nitpick, but can't this be a cache version number like `sha.update(b"1")` rather than a random hex string?

Reply:
I guess `b"1"` doesn't have enough bits set to make much of a difference in this case.  If we want to go for the version number then I guess it should be embedded in the resulting cache file.

Reply:
Maybe use something like `b"000000000000000000000000001"`?

Reply:
> I guess `b"1"` doesn't have enough bits set to make much of a difference in this case

Actually this isn't the case, a vital property of a secure hash function is collision resistance via  the avalanche effect. Even a 1-bit difference in the input will cause a significant difference in the output.

So a single byte is more than enough (the hash implementation may pad it out internally to fill a block, but we don't need to try and second-guess this - the hash designers have done it for us).

Reply:
Right, I haven't had the need to look into hash algorithm details and the like in almost two decades so I might be a bit rusty on the subject.  Anyway, I got nerd-sniped and I found this https://www.mdpi.com/2410-387X/8/3/40 so if I read the paper correctly, for SHA-256, the more bits set the better? :)

Anyway, I went with @dpgeorge's suggestion, it still encodes the same value with a few more bits set.  Nevertheless, thanks for giving me the opportunity to get re-acquainted with this stuff!

Reply:
> for SHA-256, the more bits set the better? :)

We don't need a secure hash function which achieve a particular cryptographic strength level here, we just need one which won't trivially produce collisions. If you could produce collisions by flipping a single bit or adding a single byte to a SHA input then there'd be problems with things a lot more important than this. :grin: 

> Anyway, I went with @dpgeorge's suggestion, it still encodes the same value with a few more bits set.

No worries, it's a relatively minor point anyway - one way or another we just need to change this arbitrary string if the on-disk format ever changes.